src/client/dfuse/ops/read.c

```diff
@@ -12,159 +12,160 @@
 static void
 dfuse_cb_read_complete(struct dfuse_event *ev)
 {
-	if (ev->de_ev.ev_error == 0) {
-		size_t len = ev->de_iov.iov_buf_len;
-
-		if (ev->de_len == 0)
-			DFUSE_TRA_DEBUG(ev, "Truncated read, (EOF)");
-		else if (ev->de_len != len)
-			DFUSE_TRA_DEBUG(ev,
-					"Truncated read, requested %#zx returned %#zx",
-					len, ev->de_len);
+	if (ev->de_ev.ev_error != 0) {
+		DFUSE_REPLY_ERR_RAW(ev, ev->de_req, ev->de_ev.ev_error);
+		D_GOTO(free, 0);
+	}
 
+	if (ev->de_len == 0) {
+		DFUSE_TRA_DEBUG(ev, "Truncated read, (EOF)");
 		DFUSE_REPLY_BUF(ev, ev->de_req, ev->de_iov.iov_buf, ev->de_len);
-	} else {
-		DFUSE_REPLY_ERR_RAW(ev, ev->de_req, ev->de_ev.ev_error);
+		D_GOTO(free, 0);
+	}
+
+	if (ev->de_len < ev->de_req_len)
+		DFUSE_TRA_DEBUG(ev, "Truncated read, requested %#zx returned %#zx", ev->de_req_len,
+				ev->de_len);
+	else if (ev->de_len < ev->de_iov.iov_buf_len)
+		DFUSE_TRA_DEBUG(ev, "Truncated readhead, requested %#zx returned %#zx",
+				ev->de_iov.iov_buf_len, ev->de_len);
+
+	if (ev->de_len > ev->de_req_len) {
+		struct dfuse_obj_hdl *oh;
+		int                   rc;
+
+		oh = ev->de_oh;
+		rc = pthread_mutex_trylock(&oh->doh_ie->ie_dfs->dfs_read_mutex);
+		if (rc == 0) {
+			struct dfuse_projection_info *fs_handle;
+			struct fuse_bufvec	     fb = {};
+			off_t			     ra_start;
+			off_t			     ra_len;
+
+			fs_handle      = fuse_req_userdata(ev->de_req);
+			ra_start       = ev->de_req_position + ev->de_req_len;
+			ra_len         = ev->de_len - ev->de_req_len;
+
+			fb.count       = 1;
+			fb.buf[0].mem  = ev->de_iov.iov_buf + ev->de_req_len;
+			fb.buf[0].size = ra_len;
+
+			DFUSE_TRA_DEBUG(ev, "%#zx-%#zx was readahead", ra_start,
+					ra_start + ra_len - 1);
+
+			rc = fuse_lowlevel_notify_store(fs_handle->dpi_info->di_session,
+							oh->doh_ie->ie_stat.st_ino, ra_start, &fb,
+							0);
+			if (rc == 0)
+				DFUSE_TRA_DEBUG(ev, "notify_store returned %d", rc);
+			else
+				DFUSE_TRA_INFO(ev, "notify_store returned %d", rc);
+			rc = pthread_mutex_unlock(&oh->doh_ie->ie_dfs->dfs_read_mutex);
+			if (rc != 0)
+				DFUSE_TRA_ERROR(ev, "Mutex unlock failed");
```

Comment:
It's always hard to know what to do if unlock fails, and it's very unlikely to.  I believe most of our code uses macros around the locking functions which simply log it an continue, but as this is using trylock it doesn't have access to the them.

Rather than returning an error for the I/O in hand I think it would probably be better to disable readahead if this were ever to fail.

Reply:
> better to disable readahead

Agree, since subsequent calls to trylock would fail. But I suppose this behavior existed before this PR.